math/src/main/kotlin/ktx/math/KVector.kt

```diff
@@ -0,0 +1,119 @@
+    fun isZero(margin: Float = MathUtils.FLOAT_ROUNDING_ERROR): Boolean
+
+    /** @return a vector of the same direction and a squared length of [len2] */
+    fun withLength2(len2: Float): T
+
+    /** @return Result of subtracting the given vector from this vector */
+    operator fun minus(v: T): T
+
+    /** @return Result of adding the given vector from this vector */
+    operator fun plus(v: T): T
+
+    /** @return This vector scaled by the given [scalar] */
+    operator fun times(scalar: Float): T
+
+    /** @return This vector scaled by the given [vector] */
+    operator fun times(vector: T): T
+
+    /** @return The dot product of this vector by the given [vector] */
+    fun dot(vector: T): Float
+
+    /**
+     * This method is faster than [dst] because it avoids calculating a square root. It is useful for comparisons,
+     * but not for getting exact distance, as the return value is the square of the actual distance.
+     * @return The squared distance between this and the other vector
+     */
+    fun dst2(vector: T): Float
+
+    fun lerp(target: T, alpha: Float): T
+
+    fun withRandomDirection(): T
+
+    fun isOnLine(other: T, epsilon: Float = MathUtils.FLOAT_ROUNDING_ERROR): Boolean
+
+    fun epsilonEquals(other: T, epsilon: Float): Boolean
+}
+
+/**
+ * @return The euclidean length
+ */
+val KVector<*>.len: Float get() = sqrt(len2)
+
+/**
+ * @return the unit vector of same direction or [this] if it is zero.
+ */
+val <T : KVector<T>> T.nor: T
+    get() {
+        val l2 = len2
+
+        if (l2 == 1f || l2 == 0f) return this
+
+        return withLength2(1f)
+    }
+
+/** @return Whether this vector is a unit length vector within the given margin. (no margin by default) */
+fun <T : KVector<T>> T.isUnit(margin: Float = MathUtils.FLOAT_ROUNDING_ERROR): Boolean = abs(1f - len2) < margin
+
+fun <T : KVector<T>> T.dst(v: T): Float = sqrt(dst2(v))
+fun <T : KVector<T>> T.limit(limit: Float): T = limit2(limit * limit)
+
+/** @return [this] is the [len2] is <= [limit2] and A vector with the same direction and length [limit2] otherwise */
+fun <T : KVector<T>> T.limit2(limit2: Float): T =
+        if (len2 <= limit2) this else withLength2(limit2)
+
+fun <T : KVector<T>> T.clamp(min: Float, max: Float): T = clamp2(min * min, max * max)
+
+/** @return Clamps this vector's length to given [min] and [max] values*/
+fun <T : KVector<T>> T.clamp2(min2: Float, max2: Float): T {
+    val l2 = len2
+    return when {
+        l2 < min2 -> withLength2(min2)
+        l2 > max2 -> withLength2(max2)
+        else -> this
+    }
+}
+
+fun <T : KVector<T>> T.withLength(len: Float): T = withLength2(len * len)
+
+/** @return The opposite vector of same length */
+operator fun <T : KVector<T>> T.unaryMinus(): T = times(-1f)
+
+fun <T : KVector<T>> T.interpolate(target: T, alpha: Float, interpolator: Interpolation): T =
+        lerp(target, interpolator.apply(alpha))
```

Comment:
Should I inline this kind of function?

I know many people likes inlining functions which are not more than aliases. But the gain in performance is usually between tiny and not measurable, and inline functions make implementation changes not possible without breaking binary compatibility.

Reply:
On desktop JVMs, the differences might not be noticeable, but on Android inlined functions lower the total method count and might slightly improve the performance if the function is called very often. As a rule of thumb, functions with lambda params are usually inlined in KTX.

Reply:
I haven't thought about method count indeed.

> As a rule of thumb, functions with lambda params are usually inlined in KTX

There's no lambda involved in this case. But they are *aliases*. For instance `-v` is an alias for `v * -1`. So the function could be safely inlined (reducing method count) without worrying to much about binary compatibility, or debugability (since inline function don't exist at runtime).

The argument of method counts makes me think it'd be better to inline alias-functions.

Reply:
After a quick look on the code you commented, I originally thought that interpolator is a function object and would be inlined. As I've said, I prefer that all KTX methods with lambda parameters  (and reified generics, of course) are inlined, but in other cases feel free to mark methods and properties as you consider sensible.